flow/record/base.py

```diff
@@ -724,67 +799,109 @@ def open_path(path, mode, clobber=True):
-
+    cls_stream = None
+    cls_url = None
+    adapter = None
+
+    # When a url is given, we interpret it to determine what kind of adapter we need. This piece of logic is always
+    # necessary for the RecordWriter (as it does not currently support file-like objects), and only needed for
+    # RecordReader if a url is provided.
+    if out is True or url not in ("-", "", None):
+        # Either stdout / stdin is given, or a path-like string.
+        url = str(url or "")
+        _, ext = os.path.splitext(url)
+
+        adapter_scheme = ext_to_adapter.get(ext, "stream")
+        if "://" not in url:
+            url = f"{adapter_scheme}://{url}"
+        p = urlparse(url, scheme=adapter_scheme)
+        adapter, _, sub_adapter = p.scheme.partition("+")
+
+        arg_dict = dict(parse_qsl(p.query))
+        arg_dict.update(kwargs)
+
+        cls_url = p.netloc + p.path
+        if sub_adapter:
+            cls_url = sub_adapter + "://" + cls_url
+    elif url in ("-", ""):
+        # For reading stdin, we cannot rely on an extension to know what sort of stream is incoming. Thus, we will treat
+        # it as a 'fileobj', where we can peek into the stream and try to select the appropriate adapter.
+        fileobj = getattr(sys.stdin, "buffer", sys.stdin)
+    if fileobj is not None:
+        # This record adapter has received a file-like object for record reading
+        # We just need to find the right adapter by peeking into the first few bytes.
+
+        # First, we open the stream. If the stream is compressed, open_stream will wrap it for us into a decompressor.
+        cls_stream = open_stream(fileobj, "rb")
+
+        # Now, we have a stream that will be transparently decompressed but we still do not know what adapter to use.
+        # This requires a new peek into the transparent stream. This peek will cause the stream pointer to be moved.
+        # Therefore, find_adapter_for_stream returns both a BinaryIO-supportive object that can correctly read the
+        # adjusted stream, and a string indicating the type of adapter to be used on said stream.
+        arg_dict = kwargs.copy()
+
+        # If a user did not provide a url, we have to peek into the stream to be able to determine the right adapter
+        # based on magic bytes encountered in the first few bytes of the stream.
+        if adapter is None:
+            cls_stream, adapter = find_adapter_for_stream(cls_stream)
+            if adapter is None:
+                if hasattr(cls_stream, "buffer") and cls_stream.buffer.getvalue().startswith(b"<"):
+                    raise RecordAdapterNotFound(
+                        (
+                            f"Could not find a reader for input {cls_stream.buffer.getvalue()!r}. Are you perhaps "
+                            "entering record text, rather than a recordstream? This can be fixed by using 'rdump -w -' "
+                            "to write a recordstream to stdout."
```

Comment:
```suggestion
                            "entering record text, rather than a record stream? This can be fixed by using 'rdump -w -' "
                            "to write a record stream to stdout."
```
Seems weird to put it as one word.